src/realm/object-store/c_api/query.cpp

```diff
@@ -276,13 +275,31 @@ RLM_API bool realm_query_count(const realm_query_t* query, size_t* out_count)
 RLM_API bool realm_query_find_first(realm_query_t* query, realm_value_t* out_value, bool* out_found)
 {
     return wrap_err([&]() {
-        auto key = query->query.find();
-        if (out_found)
-            *out_found = bool(key);
-        if (key && out_value) {
-            ObjLink link{query->query.get_table()->get_key(), key};
-            out_value->type = RLM_TYPE_LINK;
-            out_value->link = to_capi(link);
+        const auto& realm_query_ordering = query->get_ordering();
+        if (realm_query_ordering.size() > 0) {
+            auto orderding = util::make_bind<DescriptorOrdering>();
+            orderding->append(realm_query_ordering);
+            query->query.set_ordering(orderding);
+        }
+        if (realm_query_ordering.will_apply_sort()) {
```

Comment:
ideally this logic should be inside core's `Query::first`, and it should also check for `will_apply_distinct()` because that can also change the ordering of results